### ISSUE TITLE
Canu + minimap for correction on Guppy v3.6 reads: uncorrected has better error profile 

### ISSUE THREAD

**Issue by gringer** [marbl/canu#1715]
A pileup of uncorrected Guppy v3.6 reads produces a better guess at the true sequence compared to corrected reads. The errors are in deletion variants; SNPs are more accurately called after correction. I acknowledge that this difference may be purely in the mapper (I used mmap, rather than the default).

I've started re-assembling my January 2017 reads from _Nippostrongylus brasiliensis_, recalled using Guppy v3.6, using Canu v2.0. We've got a pretty accurate mitochondrial genome that's been previously assembled from Illumina-corrected nanopore reads (confirmed by our own Illumina cDNA reads, and by the Sanger Institute's own Illumina reads), so I've been using that as a yard stick to measure the accuracy of basecalled unmethylated reads. I use LAST with a trained alignment matrix for this mapping, because it seems to have better mapping error profiles in comparison to minimap2.

Here is a combined coverage / variant plot showing Canu-corrected reads:
![stompPlot_corrected_Nb_mtDNA](https://user-images.githubusercontent.com/856314/82058770-4cff5580-9719-11ea-8420-bc0c03754b01.png)

And here is a combined coverage / variant plot showing the uncorrected version of those same reads:
![stompPlot_Nb_uncorrected_mtDNA](https://user-images.githubusercontent.com/856314/82058621-19bcc680-9719-11ea-940e-b333cae76ca4.png)

The identified variants with frequency >40% on both the forward and reverse strand are indicated on the outer portion of the plot. Only deletion variants exist in both plots. Note that there are more identified variants for the corrected reads (11) than for the uncorrected reads (1). Based on Illumina reads that I've looked at, I believe that uncorrected variant to be a true reflection of polymorphic mitochondrial sequence, rather than an error.

This is not an issue, as such. More of an observation, just in case it's helpful for improving and/or speeding up Canu.

In case anyone wants to have a look at these reads, the uncorrected Guppy v3.6-called reads that map to the mitochondrial genome can be found [here](http://www.gringene.org/data/uncorrectedReads_vs_Nb_mtDNA.fq.gz). Coverage for the [mitochondrial genome](http://www.gringene.org/data/Nb_mtDNA_MRSR_corrected.fasta) is about 650X.

---

In order to map reads to the mitochondrial genome I use LAST. Here's an example command sequence:

    lastal -P 10 -p ~/db/last/bc.mat ~/db/fasta/nippo/Nb_mtDNA_MRSR_corrected.fasta uncorrectedReads_vs_Nb_mtDNA.fa.gz | last-split | maf-convert sam | samtools view -h --reference ~/db/fasta/nippo/Nb_mtDNA_MRSR_corrected.fasta | samtools sort > mtDNA_called_uncorrected_vs_nippo_mtDNA.bam

I have a semi-automated visualisation script that shows me coverage and variant frequencies, and determines any observed variants that are consistent on forward-mapped and reverse-mapped reads. I use this consistency check to exclude methylation-related variant signals (because it's been my experience that typically only one strand is methylated in the same place):
```
samtools view mtDNA_called_uncorrected_vs_nippo_mtDNA.bam -F 0x10 -b | samtools mpileup --reference ~/db/fasta/nippo/Nb_mtDNA_MRSR_corrected.fasta -d 100000 -Q 0 - | ~/scripts/readstomper.pl -c > stompedCounts_mtDNA_called_uncorrected_vs_nippo_mtDNA_fwd.csv
samtools view mtDNA_called_uncorrected_vs_nippo_mtDNA.bam -f 0x10 -b | samtools mpileup --reference ~/db/fasta/nippo/Nb_mtDNA_MRSR_corrected.fasta -d 100000 -Q 0 - | ~/scripts/readstomper.pl -c > stompedCounts_mtDNA_called_uncorrected_vs_nippo_mtDNA_rev.csv
~/scripts/stomp_plotter.r -f stompedCounts_fwd_4T1_ρ0SC_vs_chrM_9457.csv -r stompedCounts_rev_4T1_ρ0SC_vs_chrM_9457.csv -circular 16299 -adj 9457 -log -type png -scale -max 60000 -nodels -mindepth 2
```

---
command: 
    ~/install/canu/canu-2.0/Linux-amd64/bin/canu -nanopore-raw called_Nb_CFED_65bptrim_guppy_3.6.0.fq.gz -p   Nb_ONTCFED
_guppy360_65bpTrim -d Nb_ONTCFED_guppy360_65bpTrim genomeSize=400M corOverlapper=minimap

version: Canu 2.0

system: Debian Linux desktop
   Linux elegans 5.6.0-1-amd64 #1 SMP Debian 5.6.7-1 (2020-04-29) x86_64 GNU/Linux

**Comment by skoren** [comment]
Yes, we've seen some similar effects. I think with the increase in quality of 3.6.0 data, it makes sense to assemble them uncorrected, we're exploring similar strategies to HiFi data (compress, correct isolated errors, mask systematic errors) as well.

**Comment by harish0201** [comment]
@gringer Any such plots or outcomes from Bonito basecalled datasets?

**Comment by gringer** [comment]
Sorry, missed this question. I've done a super-accuracy recall with my mouse cDNA reads, but not yet the nippo mtDNA reads. I'll put that on my mental list to do shortly.

**Comment by gringer** [comment]
I've created an updated dataset, with a slightly more curated dataset that filters on the following conditions:
 * length >= 3kb
 * target (mitochondrial) match length >= 50% OR
 * query (read) match length >= 90%

Hopefully this will exclude most of the nuclear mitochondrial copies, if they exist. Files can be found here:

* fast5 file (20 MB): [reads_long_nippo_mtDNA.fast5](https://www.gringene.org/data/100best_guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz)
* guppy v5.1.15 called reads (1.4 MB): [guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz](https://www.gringene.org/data/guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz)
* 100 most-accurate reads (according to guppy; 463 KB): [100best_guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz](https://www.gringene.org/data/100best_guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz)

**Comment by gringer** [comment]
Now for the accuracy comparisons. I'm checking the assembled genome using LAST (with default mapping parameters, because I expect it to be very similar to the corrected assembly), and measuring the gap-compressed identity of the longest alignment to the longest assembled contig (likely less than 100% of the genome).

    # default parameters, all reads
    $ canu-2.2/bin/canu -nanopore guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz -p canu_allReads_default -d canu_allReads_default genomeSize=13k
    # GCI: 99.87%
    # default parameters, 100 most accurate reads
    $ canu-2.2/bin/canu -nanopore 100best_guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz -p canu_bestReads_default -d canu_bestReads_default genomeSize=13k
    # GCI: 99.91%
    # @skoren's recommended options, all reads
    $ canu-2.2/bin/canu -p canu_allReads_skOpts -d canu_allReads_skOpts genomeSize=13k  -trim-assemble batOptions="-eg 0.12 -sb 0.01 -dg 12 -db 12 -dr 6 -ca 1000 -cp 10" "correctedErrorRate=0.12" -pacbio-hifi guppy_5.1.15_sup_reads_long_mtDNA.fastq.gz
    # GCI: ??.??% [... this seemed to freeze at the "unitigging/3-overlapErrorAdjustment" step, so I gave up]

**Comment by gringer** [comment]
Revisiting those plots... canu does a *really* good job at read correction, essentially eliminating the random SNP errors, and most INDELs. The problem (as previously observed) is that it leaves behind some consistent errors:

<img alt="Best Raw" src="https://user-images.githubusercontent.com/856314/166237457-5f1a5461-20a6-4d10-a6dc-24f5de117896.png" width="400px" height="400px">
<img alt="Best Corrected" src="https://user-images.githubusercontent.com/856314/166237464-4e111aca-db89-4d4d-8212-76c710e95163.png"width="400px" height="400px">

<img alt="All Raw" src="https://user-images.githubusercontent.com/856314/166237473-7d0a0191-7165-451f-b4be-833edaf3bd67.png" width="400px" height="400px">
<img alt="All Corrected" src="https://user-images.githubusercontent.com/856314/166237479-23016f75-1ec2-49ad-9f8b-434cc1a92528.png" width="400px" height="400px">
